dhcpv4/bsdp/bsdp.go

```diff
@@ -0,0 +1,315 @@
+	if err != nil {
+		return nil, err
+	}
+
+	// These are vendor-specific options used to pass along BSDP information.
+	vendorOpts := []dhcpv4.Option{
+		dhcpv4.Option{
+			Code: OptionMessageType,
+			Data: []byte{MessageTypeList},
+		},
+		dhcpv4.Option{
+			Code: OptionVersion,
+			Data: Version1_1,
+		},
+	}
+
+	// If specified, replyPort MUST be a priviledged port.
+	if replyPort != 0 && replyPort != dhcpv4.ClientPort {
+		if replyPort >= 1024 {
+			return nil, fmt.Errorf("replyPort must be a privileged port (< 1024)")
+		}
+		bytes := make([]byte, 3)
+		bytes[0] = 2
+		binary.BigEndian.PutUint16(bytes[1:], replyPort)
+		d.AddOption(dhcpv4.Option{
+			Code: OptionReplyPort,
+			Data: bytes,
+		})
+	}
+	d.AddOption(dhcpv4.Option{
+		Code: dhcpv4.OptionVendorSpecificInformation,
+		Data: dhcpv4.OptionsToBytesWithoutMagicCookie(vendorOpts),
+	})
+
+	d.AddOption(dhcpv4.Option{
+		Code: dhcpv4.OptionParameterRequestList,
+		Data: []byte{
+			dhcpv4.OptionVendorSpecificInformation,
+			dhcpv4.OptionClassIdentifier,
+		},
+	})
+
+	u16 := make([]byte, 2)
+	binary.BigEndian.PutUint16(u16, 1500)
+	d.AddOption(dhcpv4.Option{
+		Code: dhcpv4.OptionMaximumDHCPMessageSize,
+		Data: u16,
+	})
+
+	vendorClassID, err := makeVendorClassIdentifier()
+	if err != nil {
+		return nil, err
+	}
+	d.AddOption(dhcpv4.Option{
+		Code: dhcpv4.OptionClassIdentifier,
+		Data: []byte(vendorClassID),
+	})
+
+	d.AddOption(dhcpv4.Option{Code: dhcpv4.OptionEnd})
+	return d, nil
+}
+
+// InformSelectForAck constructs an INFORM[SELECT] packet given an ACK to the
+// previously-sent INFORM[LIST] with Config config.
+func InformSelectForAck(ack dhcpv4.DHCPv4, replyPort uint16, selectedImage BootImage) (*dhcpv4.DHCPv4, error) {
+	d, err := dhcpv4.New()
+	if err != nil {
+		return nil, err
+	}
+	d.SetOpcode(dhcpv4.OpcodeBootRequest)
+	d.SetHwType(ack.HwType())
+	d.SetHwAddrLen(ack.HwAddrLen())
+	clientHwAddr := ack.ClientHwAddr()
+	d.SetClientHwAddr(clientHwAddr[:])
+	d.SetTransactionID(ack.TransactionID())
+	if ack.IsBroadcast() {
+		d.SetBroadcast()
+	} else {
+		d.SetUnicast()
+	}
+
+	// Data for OptionSelectedBootImageID
+	vendorOpts := []dhcpv4.Option{
+		dhcpv4.Option{
+			Code: OptionMessageType,
+			Data: []byte{MessageTypeSelect},
+		},
+		dhcpv4.Option{
+			Code: OptionVersion,
+			Data: Version1_1,
+		},
+		dhcpv4.Option{
+			Code: OptionSelectedBootImageID,
+			Data: selectedImage.ID.ToBytes(),
+		},
+	}
+
+	// Find server IP address
+	var serverIP net.IP
+	// TODO replace this loop with `ack.GetOneOption(OptionBootImageList)`
+	for _, opt := range ack.Options() {
+		if opt.Code == dhcpv4.OptionServerIdentifier {
+			serverIP = net.IP(opt.Data)
+		}
+	}
+	if serverIP.To4() == nil {
+		return nil, fmt.Errorf("could not parse server identifier from ACK")
+	}
+	vendorOpts = append(vendorOpts, dhcpv4.Option{
+		Code: OptionServerIdentifier,
+		Data: serverIP,
+	})
+
+	// Validate replyPort if requested.
+	if replyPort != 0 && replyPort != dhcpv4.ClientPort {
```

Comment:
this block is duplicated from `NewInformListForInterface`, can you make it a private function?